### PR TITLE
Iep 651 restructure insolvency collection for 600

### PR DESCRIPTION
**Pull request by Jideboris**
Please note mongo helper was introduced but not fully used now.
Data model would need to be cleaned up on separate ticket.

### CODE REVIEW

File: dao/mongo.go [hunk 1]
```diff
@@ -179,124 +232,97 @@ func (m *MongoService) CreatePractitionerResourceForInsolvencyCase(dao *models.P
...
+func (m *MongoService) GetPractitionersByIds(practitionerIDs []string, transactionID string) ([]models.PractitionerResourceDto, error) {
+	var insolvencyResource []models.PractitionerResourceDto
+	collection := m.db.Collection(PractitionerCollectionName)
 
 	filter := bson.M{
-		"transaction_id":        transactionID,
-		"data.practitioners.id": practitionerID,
+		"data.id": bson.M{
+			"$in": bson.A{practitionerIDs},
+		},
```
Comment (gshivach) [inline dao/mongo.go:309]: Shouldn't this request get practitioner specific to the transaction?

File: dao/service.go [hunk 2]
```diff
@@ -7,30 +7,42 @@ import (
...
+	// GetInsolvencyPractitionerByTransactionID will retrieve a list of persisted practitioners
+	GetInsolvencyPractitionerByTransactionID(transactionID string) (string, error)
+
+	// GetPractitionerAppointment will retrieve a practitioner appointment
+	GetPractitionerAppointment(practitionerID string, transactionID string) (*models.AppointmentResourceDao, error)
 
-	// GetPractitionerResource will retrieve a practitioner from the insolvency resource
-	GetPractitionerResource(practitionerID string, transactionID string) (models.PractitionerResourceDao, error)
+	// GetPractitionersByIds will retrieve a practitioner from the insolvency resource
+	GetPractitionersByIds(practitionerIDs []string, transactionID string) ([]models.PractitionerResourceDto, error)
```
Comment (nevans4-ch) [inline dao/service.go:33]: Is this singular or plural? (comment says 'a practitioner', function name says Practitioners)

File: dao/mongo.go [hunk 3]
```diff
@@ -129,7 +132,37 @@ func (m *MongoService) GetInsolvencyResource(transactionID string) (models.Insol
 		return models.InsolvencyResourceDao{}, fmt.Errorf("there was a problem handling your request for transaction [%s]", transactionID)
 	}
 
-	return insolvencyResource, nil
+	// convert insolvencyResourceDto to insolvencyResourceDao
+	insolvencyResourceDao := transformers.InsolvencyResourceDtoToInsolvencyResourceDao(insolvencyResource)
+
+	return insolvencyResourceDao, nil
+}
+func (m *MongoService) GetPractitionerAppointment(practitionerID string, transactionID string) (*models.AppointmentResourceDao, error) {
```
Comment (nevans4-ch) [inline dao/mongo.go:140]: Are we passing in the transactionID only for logging purposes here?

**Review by nevans4-ch — COMMENTED** [review]
Comments added inline - mostly to do with function naming & commenting to improve readability.  I would also like to see some additional checks (on field values etc) added to tests to increase our confidence during these changes, as some are currently doing quite shallow checks.  I also noticed that a few tests were deleted (from mongo_driver_test.go & insolvency_resource_test.go) - are we sure that coverage has not decreased?

**Review by nevans4-ch — COMMENTED** [review]
A couple of additional queries, sorry!

File: dao/mongo.go (hunk 1 above)
Comment (Jideboris) [reply dao/mongo.go:274]: discussed

File: models/data_entities.go [hunk 4]
```diff
@@ -49,21 +71,22 @@ type PractitionerResourceDao struct {
...
 }
 
 // AppointmentResourceDao contains the appointment data for a practitioner
 type AppointmentResourceDao struct {
-	AppointedOn string                      `bson:"appointed_on,omitempty"`
-	MadeBy      string                      `bson:"made_by,omitempty"`
-	Links       AppointmentResourceLinksDao `bson:"links,omitempty"`
-	Etag        string                      `bson:"etag"`
-	Kind        string                      `bson:"kind"`
+	PractitionerID string                      `bson:"id"`
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply models/data_entities.go]: Fixed

File: utils/converter.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package utils
+
+import (
+	"encoding/json"
+)
+
+// ConvertToMap is a helper function to convert string to map object
+func ConvertStringToMap(mapString string) (map[string]string, []string, error) {
+	var mapResource map[string]string
+	var stringPractitonerIdsArray []string
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply utils/converter.go]: Fixed

File: models/response_entities.go [hunk 6]
```diff
@@ -48,9 +48,11 @@ type CreatedPractitionerLinksResource struct {
 
 // AppointedPractitionerResource contains the details of an appointed practitioner
 type AppointedPractitionerResource struct {
-	AppointedOn string                             `json:"appointed_on"`
-	MadeBy      string                             `json:"made_by"`
-	Links       AppointedPractitionerLinksResource `json:"links"`
+	AppointedOn string                      `json:"appointed_on"`
+	MadeBy      string                      `json:"made_by"`
+	Links       AppointmentResourceLinksDao `json:"links"`
+	Etag           string                  `json:"etag"`
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply models/response_entities.go:54]: Discussed

File: utils/converter.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package utils
+
+import (
+	"encoding/json"
+)
+
+// ConvertToMap is a helper function to convert string to map object
+func ConvertStringToMap(mapString string) (map[string]string, []string, error) {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply utils/converter.go]: Fixed

File: handlers/practitioner_resource.go [hunk 8]
```diff
@@ -385,30 +406,42 @@ func HandleDeletePractitionerAppointment(svc dao.Service) http.Handler {
...
-			utils.WriteJSONWithStatus(w, req, m, statusCode)
+			logError(w, req, statusCode, "error", []error{err})
 			return
 		}
 
 		w.WriteHeader(statusCode)
 	})
 }
 
+func logError(w http.ResponseWriter, req *http.Request, status int, kind string, errs []error) {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply handlers/practitioner_resource.go]: Fixed

File: handlers/practitioner_resource_test.go [hunk 9]
```diff
@@ -656,17 +702,32 @@ func serveGetPractitionerResourceRequest(service dao.Service, tranIDSet bool, pr
 	return res
 }
 
-func TestUnitHandleGetPractitionerResource(t *testing.T) {
+func TestUnitHandleGetPractitionersByIds(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply handlers/practitioner_resource_test.go:714]: Fixed

File: transformers/insolvency_resource.go [hunk 10]
```diff
@@ -3,59 +3,75 @@ package transformers
...
 	"github.com/companieshouse/insolvency-api/models"
 	"github.com/companieshouse/insolvency-api/utils"
 )
 
-// InsolvencyResourceRequestToDB will take the input request from the REST call and transform it to a dao ready for
-// insertion into the database
-func InsolvencyResourceRequestToDB(req *models.InsolvencyRequest, transactionID string, helperService utils.HelperService) *models.InsolvencyResourceDao {
+// InsolvencyResourceDaoToCreatedResponse will transform an insolvency resource dao that has successfully been created into
+// a http response entity
+func InsolvencyResourceDtoToInsolvencyResourceDao(insolvencyResourceDto models.InsolvencyResourceDto) models.InsolvencyResourceDao {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply transformers/insolvency_resource.go:13]: Discussed

File: dao/mongo.go [hunk 11]
```diff
@@ -164,10 +197,30 @@ func (m *MongoService) CreateAppointmentResource(appointmentResourceDao *models.
...
+	}
+
+	return http.StatusNoContent, nil
+}
+
 // CreatePractitionerResourceForInsolvencyCase stores an incoming practitioner to the list of practitioners for the insolvency case
 // with the specified transactionID
-func (m *MongoService) CreatePractitionerResourceForInsolvencyCase(dao *models.PractitionerResourceDao, transactionID string) (int, error) {
-	var insolvencyResource models.InsolvencyResourceDao
+func (m *MongoService) CreatePractitionerResourceForInsolvencyCase(transactionID string) (map[string]string, int, error) {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply dao/mongo.go]: Fixed

File: dao/service.go (hunk 2 above)
Comment (Jideboris) [reply dao/service.go]: Fixed

File: dao/mongo.go (hunk 3 above)
Comment (Jideboris) [reply dao/mongo.go:141]: Discussed to be kept like that for now

File: service/practitioner_service.go [hunk 12]
```diff
@@ -71,17 +71,31 @@ func ValidatePractitionerDetails(svc dao.Service, transactionID string, practiti
 }
 
 // ValidateAppointmentDetails checks that the incoming appointment details are valid
-func ValidateAppointmentDetails(svc dao.Service, appointment models.PractitionerAppointment, transactionID string, practitionerID string, req *http.Request) (string, error) {
+func ValidateAppointmentDetails(svc dao.Service, appointment models.PractitionerAppointment, transactionID string, practitionerID string, req *http.Request) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply service/practitioner_service.go:74]: Fixed thanks

File: handlers/insolvency_resource.go [hunk 13]
```diff
@@ -20,6 +20,13 @@ import (
 func HandleCreateInsolvencyResource(svc dao.Service, helperService utils.HelperService) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 
+		// generate etag for request
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply handlers/insolvency_resource.go:23]: Discussed

File: dao/mongo.go [hunk 14]
```diff
@@ -214,119 +213,85 @@ func (m *MongoService) UpdateInsolvencyPractitioners(practitionersResource model
...
 	if len(insolvencyResource.Data.Practitioners) > 0 {
 		err = json.Unmarshal([]byte(insolvencyResource.Data.Practitioners), &practitionersData)
 		if err != nil {
 			log.Error(err)
-			return nil, http.StatusInternalServerError, fmt.Errorf("there was a problem handling json Unmarshalling %s", transactionID)
+			return nil, "", fmt.Errorf("there was a problem handling json Unmarshalling %s", transactionID)
 		}
 	}
 
 	// Check if there are already 5 practitioners in database
```
Comment (nevans4-ch) [inline dao/mongo.go:255]: The check for more than 5 practitioners doesn't fit well here now that it is a get not create operation - should it be moved up to the handler?

File: models/data_entities.go [hunk 15]
```diff
@@ -71,22 +70,23 @@ type PractitionerResourceDao struct {
 
 // PractitionerResourceDto contains the data for the practitioner resource in Mongo
 type PractitionerResourceDto struct {
+	ID   string                  `bson:"practitioner_id"`
 	Data PractitionerResourceDao `bson:"data"`
 }
 
 // AppointmentResourceDto contains the data for the appointment resource in Mongo
 type AppointmentResourceDto struct {
+	ID   string                 `bson:"practitioner_id"`
```
Comment (nevans4-ch) [inline models/data_entities.go:79]: Could the struct field name also be called PractitionerId for clarity please?

File: dao/mongo.go [hunk 16]
```diff
@@ -214,119 +213,85 @@ func (m *MongoService) UpdateInsolvencyPractitioners(practitionersResource model
 
 	return http.StatusNoContent, nil
 }
+ 
+// GetInsolvencyPractitionerByTransactionID gets a list of all practitioners for an insolvency case with the specified transactionID
+func (m *MongoService) GetInsolvencyPractitionerByTransactionID(transactionID string) (map[string]string, string, error) {
```
Comment (nevans4-ch) [inline dao/mongo.go:218]: The return types here don't seem to fit with the dao package - all other get operations return just one or more Dao/Dto objects, in line with the package's purpose as an interface for the persistence layer.  Are we doing conversions here that would be better handled by the calling functions?

**Review by nevans4-ch — COMMENTED** [review]
Inline comments added on follow-up commits.
Please note that there are also 2 of my prev comments remaining open (may need to click 'Load more...' in hidden conversations section to see them, as GitHub had collapsed the middle section...)

File: dao/mongo.go [hunk 17]
```diff
@@ -166,13 +168,11 @@ func (m *MongoService) GetPractitionerAppointment(practitionerID string, transac
 }
 
 // CreatePractitionerResource stores an incoming practitioner to the practitioners collection
-func (m *MongoService) CreatePractitionerResource(practitionerResourceDao *models.PractitionerResourceDao, transactionID string) (int, error) {
-	var practitionerData models.PractitionerResourceDto
+func (m *MongoService) CreatePractitionerResource(practitionerResourceDao *models.PractitionerResourceDto, transactionID string) (int, error) {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply dao/mongo.go:171]: The way the models are to be sorted out in a ticket.However, I have updated as requested now.Thanks

File: dao/mongo.go (hunk 14 above)
Comment (Jideboris) [reply dao/mongo.go:255]: Agreed, oversight.Done.Thanks

File: models/data_entities.go (hunk 15 above)
Comment (Jideboris) [reply models/data_entities.go:79]: BSON already given the definition of what the ID represent, however changed as requested

File: dao/mongo.go (hunk 16 above)
Comment (Jideboris) [reply dao/mongo.go:218]: Good point resolved.Thanks

File: handlers/practitioner_resource.go [hunk 18]
```diff
@@ -76,11 +73,46 @@ func HandleCreatePractitionersResource(svc dao.Service, helperService utils.Help
 		practitionerDao.Kind = "insolvency#practitioner"
 
 		// Store practitioners resource in practitioners collection
-		statusCode, err := svc.CreatePractitionerResourceForInsolvencyCase(practitionerDao, transactionID)
+		practionersResource, statusCode, err := svc.CreatePractitionerResourceForInsolvencyCase(transactionID)
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply handlers/practitioner_resource.go]: Fixed

File: handlers/practitioner_resource_test.go [hunk 19]
```diff
@@ -555,17 +584,35 @@ func serveGetPractitionerResourcesRequest(service dao.Service, tranIDSet bool) *
 	return res
 }
 
-func TestUnitHandleGetPractitionerResources(t *testing.T) {
+func TestUnitHandleGetPractitionersByIdss(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Jideboris) [reply handlers/practitioner_resource_test.go]: Fixed

**Review by nevans4-ch — APPROVED** [review]

**Pull request merged**